### ISSUE TITLE
Page.insertImage fails with keep_proportion=False

### ISSUE THREAD

**Issue by epitavy** [pymupdf/PyMuPDF#728]
My config:
 - Operating system: Linux with Ubuntu 20.04.1 LTS
 - Python version: 3.8.5
 - PyMuPDF version: 1.18.4, installed with pip 20.0.2

I have an application that generates QRcode, which are available under stream byte. I don't want to save these on disk. I want to insert these QR codes on an single pdf pages and I use the `insertImage` of the `Page` class. My code is the following:
```python
>>> doc = fitz.Document()
>>> doc.newPage()
>>> page = doc.loadPage(0)
>>> type(img)
<class 'PIL.Image.Image'>
>>> img_stream = img.tobytes()
>>> type(img_stream)
<class 'bytes'>
>>> rect = fitz.Rect(0, 0, 290, 290)
>>> page.insertImage(rect, stream=img_stream)
File "/home/user/.venv/lib/python3.8/site-packages/fitz/utils.py", line 331, in insertImage
    w, h = img_prof["width"], img_prof["height"]
TypeError: 'NoneType' object is not subscriptable
>>> page.insertImage(rect, stream=img_stream, keep_proportion=False)
File "/home/user/.venv/lib/python3.8/site-packages/fitz/utils.py", line 364, in insertImage
    xref = doc.InsertedImages.get(digest, 0)  # reuse any previously inserted image
UnboundLocalError: local variable 'digest' referenced before assignment
```
I expected some explicit error message or the function to work and add my image in the page.

I first thought that I got errors because the byte stream is not exactly as expected but I looked a bit in the code and it seems that some checks are missing.
It appears to me that `page.insertImage(rect, stream=img_stream)` fails because `TOOLS.image_profile(stream, keep_image=True)` in utils.py returns `None`.
This piece of code is not called whith keep_proportion=False, so I made a second attempt with `page.insertImage(rect, stream=img_stream, keep_proportion=False)` and this time it fails because `digest` in only defined when keep_proportion=True.

**Comment by JorjMcKie** [comment]
You are right, I should check, that ``rect`` is the only positional parameter.
Your code should work if you use ``stream=img_stream``.

**Comment by epitavy** [comment]
Sorry, I made some typos in my issue, I effectively run PyMuPDF version 1.18.4 and not 1.18.14, and I used `stream=img_stream` instead of a positional parameter. I edited the post. I tested to pass the parameter as a positional parameter and nothing happens, except that the byte stream is displayed on the console.

**Comment by JorjMcKie** [comment]
No problem - thanks for submitting this anyway.
But your code works if you use the ``stream`` keyword, doesn't it?

**Comment by epitavy** [comment]
No, it doesn't work. I tried to save the file and then open it as a byte stream, to fit the example code but I have the same errors.

**Comment by JorjMcKie** [comment]
Ah, please look in the Pillow documentation - you should not use Image.tobytes():

Quote:
--------------------------------------------------------------------------------------------
Image.tobytes(encoder_name=’raw’, *args)
Return image as a bytes object.
Warning: This method returns the raw image data from the internal storage. For compressed image data (e.g. PNG, JPEG) use save(), with a BytesIO parameter for in-memory data.
--------------------------------------------------------------------------------------------

So you should do something like this:

bio = io.BytesIO()
Image.save(bio, “png”)
page.insertImage(rect, stream=bio,…)

**Comment by epitavy** [comment]
Thanks for you answer, it works now!
However the issue is not solved and the variable `digest` is not defined when the parameter `keep_proportion=False` is used.

**Comment by JorjMcKie** [comment]
> However the issue is not solved and the variable digest is not defined when the parameter keep_proportion=False is used.

That's correct. I found the error. Easy to fix.
Do you depend on this? I can either send you a hot fix with a new version of ``utils.py``, whichyou would simple have to copy over the one of your installation.
Or produce a pre-version wheel, which you could download ...
Please drop a note.

**Comment by epitavy** [comment]
No it is fine, I don't depend on it. Your fix with BytesIO was perfect.

**Comment by JorjMcKie** [comment]
Solved by v1.18.5 being uploaded now.
